extension_templates/forecasting.py

```diff
@@ -0,0 +1,244 @@
+# -*- coding: utf-8 -*-
+"""
+Extension template for forecasters.
+
+How to use this:
+- this is meant as a "fill in" template for easy extension
+- do NOT import this file directly - it will break
+- work through all the "todo" comments below
+- fill in code for mandatory methods, and optionally for optional methods
+- you can add more private methods, but do not override BaseEstimator's private methods
+    an easy way to be safe is to prefix your methods with "_custom"
+- change docstrings for functions and the file
+- ensure interface compatibility by testing forecasting/tests/test_all_forecasters
+        and forecasting/tests/test_sktime_forecasters
+- once complete: use as a local library, or contribute to sktime via PR
+
+Mandatory implements:
+    fitting         - _fit(self, y, X=None, fh=None)
+    forecasting     - _predict(self, fh=None, X=None, return_pred_int=False,
+                               alpha=DEFAULT_ALPHA)
+
+Optional implements:
+    updating        - _update(self, y, X=None, update_params=True):
+    fitted parameter inspection - get_fitted_params()
+
+State:
+    fitted model/strategy   - by convention, any attributes ending in "_"
+    fitted state flag       - check_is_fitted()
+
+copyright: sktime developers, BSD-3-Clause License (see LICENSE file)
+"""
+
+from sktime.forecasting.base import BaseForecaster
+from sktime.forecasting.base._base import DEFAULT_ALPHA
+
+# todo: add any necessary imports here
+
+
+class MyForecaster(BaseForecaster):
+    """Custom forecaster. todo: write docstring.
+
+    todo: describe your custom forecaster here
+
+    Hyper-parameters
+    ----------------
+    parama : int
+        descriptive explanation of parama
+    paramb : string, optional (default='default')
+        descriptive explanation of paramb
+    paramc : boolean, optional (default= whether paramb is not the default)
+        descriptive explanation of paramc
+    and so on
+
+    Components
+    ----------
+    est : sktime.estimator, BaseEstimator descendant
+        descriptive explanation of est
+    est2: another estimator
+        descriptive explanation of est2
+    and so on
+    """
+
+    # todo: add any hyper-parameters and components to constructor
+    def __init__(self, parama, paramb="default", paramc=None):
+
+        # todo: write any hyper-parameters and components to self
+        self.parama = parama
+        self.paramb = paramb
+        self.paramc = paramc
+
+        # todo: initialize None parameters, where necessary
+        if paramc is None:
```

Comment:
hyper-parameters should be stored as they are and interpreted in fit

Reply:
mh, but isn't it common practice to move complex defaults into the constructor? E.g., if the default of `estimator_param` is some pipeline construct, you would set the argdefault to `None` and replace it with the pipeline construct initialized in the constructor if `estimator_param is None`?

Reply:
is that a sklearn convention?

Reply:
I believe so, I think otherwise settings hyper-parameters in `estimator.set_params()` will be treated differently than in the constructor. The constructor should be largely free of any logic on hyper-parameters.

Reply:
not sure whether I see this - I know that get/set params look at the `__init__` signature, and there should be no kwargs. Otherwise, I don't see how initialization would affect the `BaseEstimator` functionality - would you mind explaining?

The principle by which I was going is that initialization (including of default parameters) should happen in ... the constructor, also named `__init__` for that reason.

Reply:
Take a look at scikit-learn's dev guide where they state this: https://scikit-learn.org/stable/developers/develop.html#apis-of-scikit-learn-objects

Reply:
I think the (undocumented) exception is estimators (not parameters!) passed to `__init__`, which can have validation and/or init logic? For example the `Pipeline` validates its `steps` in `__init__`, see https://github.com/scikit-learn/scikit-learn/blob/main/sklearn/pipeline.py

For initialization of default estimators, I think there's no precedent for that in scikit-learn. But, in extension, I'd also think it's fine, and even necessary. Because `get/set_params` looks at the composite estimator, and wouldn't "see" the parameters of the default estimator unless it's already set in `__init__`.

Reply:
Yes, interesting, they do call it again though in [`_fit`](https://github.com/scikit-learn/scikit-learn/blob/9a13bdfaf1a47188d2e1262f0308f317e6662e8d/sklearn/pipeline.py#L278)

Reply:
So I suggest to remove the parameter initialisation from the template.

Reply:
ah, I see - yes, will do.

Reply:
I've updated the extension template now along these lines, have a look.

Reply:
looks good now